### PR TITLE
feature: create a Trust on First Use example

### LINKED ISSUE #9102: Provide some middle ground between InSecure and WiFiClientSecure
### Related area

SSL Security

### Hardware specification

any

### Is your feature request related to a problem?

Increasingly the WifiClientInsecure example seems to be making it into actual products on the market. Which is not ideal.

### Describe the solution you'd like

At the very least provide people with some middle ground; i.e. when their deployment setting does not allow for a hardcode CA cert (e.g. when the server is not known ahead of time).

### Describe alternatives you've considered

An elaborate example with a web-portal that shows how to upload/enter a certificate; but this moves some of the hard complexity to endusers which then require a lot of skills.

### Additional context

_No response_

### I have checked existing list of Feature requests and the Contribution Guide

- [X] I confirm I have checked existing list of Feature requests and Contribution Guide.

### PR DESCRIPTION
**Pull request by dirkx**
create a Trust on First Use example to quell the increasingly ommon copy & paste of the insecure approach making it to production

This addresses #9102 

## Description of Change

Add an example. No impact. Put a warning in the existing NonSecure example (little impact)

## Tests scenarios

ESP32

## Related links
Would close #9102

### CODE REVIEW

**Comment by me-no-dev** [comment]
you need to add `.skip.esp32h2` files to both sketch folders, so they are not compiled for H2, which does not have WiFi

**Comment by dirkx** [comment]
Ah thanks - I was already wondering. Will fix.

**Comment by VojtechBartoska** [comment]
@lucasssvaz & @P-R-O-C-H-Y Please help with review when this is ready.

File: libraries/WiFiClientSecure/examples/WiFiClientTrustOnFirstUse/WiFiClientTrustOnFirstUse.ino [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+    }
+  }
+  Serial.println("\n\n-- Payload ended.");
+  client.stop();
+  return true;
+}
+
+void loop() {
+  delay(-1);
+}
```
Comment (lucasssvaz) [inline libraries/WiFiClientSecure/examples/WiFiClientTrustOnFirstUse/WiFiClientTrustOnFirstUse.ino:266]: No need for this. It's just an example

```suggestion
void loop() {}
```

**Comment by dirkx** [comment]
I am not quite seeing that on an ESP32-S2 and DEBUG set to VERBOSE. Anything extra I need to set ?

My first guess is that the server closes the connection:

18:08:59.617 -> [  2947][E][WiFiClientSecure.cpp:316] available(): Closing connection on failed avail check
18:08:59.617 -> [  2949][V][ssl_client.cpp:337] stop_ssl_socket(): Cleaning SSL connection.

And that 'we' then close the connection again:

18:09:00.913 -> [  4232][V][ssl_client.cpp:337] stop_ssl_socket(): Cleaning SSL connection.

That could be solved in

     void WiFiClientSecure::stop()
     {
         if (sslclient->socket >= 0) {
             close(sslclient->socket);
            _lastWriteTimeout = 0;
.....
          }
          stop_ssl_socket(sslclient, _CA_cert, _cert, _private_key);
     }

by moving that stop_ssl_socket() into the IF statement. But then we're changing the API a little bit. Not sure if that is wise.

Dw

8:08:58.506 -> Connected to Linate
18:08:58.506 -> Trying to connect to a server; using TOFU details from the eeprom
18:08:58.576 -> [  1890][V][ssl_client.cpp:60] start_ssl_client(): Free internal heap before TLS 146408
18:08:58.576 -> [  1890][V][ssl_client.cpp:66] start_ssl_client(): Starting socket
18:08:58.714 -> [  2044][V][ssl_client.cpp:144] start_ssl_client(): Seeding the random number generator
18:08:58.714 -> [  2046][V][ssl_client.cpp:153] start_ssl_client(): Setting up the SSL/TLS structure...
18:08:58.714 -> [  2049][D][ssl_client.cpp:174] start_ssl_client(): WARNING: Skipping SSL Verification. INSECURE!
18:08:58.748 -> [  2057][V][ssl_client.cpp:263] start_ssl_client(): Setting hostname for TLS session...
18:08:58.748 -> [  2065][V][ssl_client.cpp:285] ssl_starttls_handshake(): Performing the SSL/TLS handshake...
18:08:59.380 -> [  2713][V][ssl_client.cpp:306] ssl_starttls_handshake(): Verifying peer X.509 certificate...
18:08:59.380 -> [  2713][V][ssl_client.cpp:314] ssl_starttls_handshake(): Certificate verified.
18:08:59.415 -> [  2717][V][ssl_client.cpp:330] ssl_starttls_handshake(): Free internal heap after TLS 102744
18:08:59.415 -> All well - you are talking to the same server as
18:08:59.415 -> when you set up TOFU. So we can now do a GET.
18:08:59.415 -> 
18:08:59.415 -> 
18:08:59.583 -> HTTP/1.0 200 OK
Access-Control-Allow-Origin: *
Content-Length: 3299
Content-Type: application/json
Strict-Transport-Security: max-age=631138519; includeSubdomains; preload
Vary: Accept-Encoding
Date: Thu, 18 Jan 2024 17:08:59 GMT

-- headers received. Payload follows
18:08:59.583 -> 
18:08:59.583 -> 
18:08:59.617 -> [  2947][E][WiFiClientSecure.cpp:316] available(): Closing connection on failed avail check
18:08:59.617 -> [  2949][V][ssl_client.cpp:337] stop_ssl_socket(): Cleaning SSL connection.
18:09:00.611 -> {"given_cipher_suites":["TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384","TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384","TLS_DHE_RSA_WITH_AES_256_GCM_SHA384","TLS_ECDHE_ECDSA_WITH_AES_256_CCM","TLS_DHE_RSA_WITH_AES_25....splitting":false,"insecure_cipher_suites":{},"tls_version":"TLS 1.2","rating":"Probably Okay"}
18:09:00.913 -> 
18:09:00.913 -> -- Payload ended.
18:09:00.913 -> ALL OK

**Comment by lucasssvaz** [comment]
I tested on a ESP32 in the verbose mode. Idk if that should make a difference

**Comment by dirkx** [comment]
Ok - will try an ESP32 rather than an ESP32-S2. Meanwhile - I did observe that, with the exception of setting the value of ssl_client->socket; al other socket work is delegated to the ssl_client.c world.

So below diff does make things a bit cleaner; and at least surpresses the double closing when a socket normally closes and the user does a stop later on (again). If we assume that nothing calls stuff in ssl_client - then the API is not changing.

<code>
       diff --git a/libraries/WiFiClientSecure/src/WiFiClientSecure.cpp b/libraries/WiFiClientSecure/src/WiFiClientSecure.cpp
       index ff2dedf5..0339e642 100644
       --- a/libraries/WiFiClientSecure/src/WiFiClientSecure.cpp
       +++ b/libraries/WiFiClientSecure/src/WiFiClientSecure.cpp
       @@ -20,8 +20,6 @@
          #include "WiFiClientSecure.h"
        #include "esp_crt_bundle.h"
       -#include <lwip/sockets.h>
       -#include <lwip/netdb.h>
        #include <errno.h>
          #undef connect
       @@ -91,15 +89,11 @@ WiFiClientSecure &WiFiClientSecure::operator=(const WiFiClientSecure &other)
          void WiFiClientSecure::stop()
        {
       -    if (sslclient->socket >= 0) {
       -        close(sslclient->socket);
       -        sslclient->socket = -1;
       -        _connected = false;
       -        _peek = -1;
       -        _lastReadTimeout = 0;
       -        _lastWriteTimeout = 0;
       -    }
            stop_ssl_socket(sslclient, _CA_cert, _cert, _private_key);
       +    _connected = false;
       +    _peek = -1;
       +    _lastReadTimeout = 0;
       +    _lastWriteTimeout = 0;
        }
          int WiFiClientSecure::connect(IPAddress ip, uint16_t port)
</code>

**Comment by dirkx** [comment]
Actually - looking at this a bit longer - I think you unearthed a much older bug due to that logic (I removed in above patch).

If there is an error and ssl_client level class close the connection; they set ->socket to -1.

But the WiFiSecureClient layer above it checks on _connected mostly (e.g. in read, write, etc).

But _connected may not get unset - as WiFiSecureClient gate the set to false that on socket not being -1. Which can have happened in the layer below of ssl_client. 

So I guess my patch may actually solve that bug - unrelated to any of the TOFU / etc work.

With kind regards,

Dw.

**Comment by dirkx** [comment]
Apologies - will do - but paid customers intruding :)

**Review by lucasssvaz — APPROVED** [review]
Bug found is unrelated to the example added. I'll open a issue to track it

**Pull request merged**